### PR TITLE
UI tests table

### PR DESCRIPTION
**Pull request by ahuzmeza**
Implemented 'Tests' UI screen containing:
-- segmented control via stack view
-- one tableview with 3 custom cell types
-- chapters/ final/ passed cell

### CODE REVIEW

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsCustomCells/FinalTableViewCell.swift [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+//  Internship2022TeamSzabi
+//
+//  Created by Alin Huzmezan on 8/5/22.
+//
+
+import UIKit
+
+class FinalTableViewCell: UITableViewCell {
+    @IBOutlet private weak var cellView: UIView!
+    @IBOutlet weak var titleLabel: UILabel!
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsCustomCells/FinalTableViewCell.swift]: private?

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsCustomCells/FinalTableViewCell.swift [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+    func updateCellView(finalModel: FinalTestModel) {
+        titleLabel.text = finalModel.title
+        subTitleLabel.text = finalModel.subTitle
+    }
+
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        // Initialization code
+        cellView.layer.cornerRadius = 20
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsCustomCells/FinalTableViewCell.swift:25]: maybe move it in a new method, setupView or something

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsCustomCells/FinalTableViewCell.swift [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import UIKit
+
+class FinalTableViewCell: UITableViewCell {
+    @IBOutlet private weak var cellView: UIView!
+    @IBOutlet weak var titleLabel: UILabel!
+    @IBOutlet weak var subTitleLabel: UILabel!
+
+    static let identifier = "finalCell"
+
+    func updateCellView(finalModel: FinalTestModel) {
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsCustomCells/FinalTableViewCell.swift:17]: maybe is a preference of mine, but I would like to have in classes the following order: outlets / overrides / other methods

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsCustomCells/FinalTableViewCell.swift [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+//
+
+import UIKit
+
+class FinalTableViewCell: UITableViewCell {
+    @IBOutlet private weak var cellView: UIView!
+    @IBOutlet weak var titleLabel: UILabel!
+    @IBOutlet weak var subTitleLabel: UILabel!
+
+    static let identifier = "finalCell"
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsCustomCells/FinalTableViewCell.swift:15]: you can create a struct / enum of constants, also for corner radius / margins

**Review by IuliaRAnisoi — CHANGES_REQUESTED** [review]

**Review by IuliaRAnisoi — CHANGES_REQUESTED** [review]

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsCustomCells/ChaptersTableViewCell.swift [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        grade.isHidden = true
+        arrow.isHidden = false
+    }
+
+    func updateCellViewforPassed(/* PassedModel */) {
+        grade.isHidden = false
+        arrow.isHidden = true
+    }
+
+    override func awakeFromNib() {
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsCustomCells/ChaptersTableViewCell.swift:30]: same here with he order

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsViewController.swift [hunk 6]
```diff
@@ -155,32 +155,23 @@ extension TestsViewController: UITableViewDataSource, UITableViewDelegate {
...
-                let cell = tableView.dequeueReusableCell(
+            guard let cell = tableView.dequeueReusableCell(
                     withIdentifier: ChaptersTableViewCell.identifier, for: indexPath)
                     as? ChaptersTableViewCell
-            else {
-                return UITableViewCell()
-            }
+            else { return UITableViewCell() }
             cell.updateCellViewforPassed(/* PassedModel @ indexPath */)
             return cell
```
[the remark this answers is not in the document]
Comment (ahuzmeza) [reply Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsViewController.swift:176]: SwiftLint line length violation. I'll separate guard into 3 separate lines

**Review by IuliaRAnisoi — APPROVED** [review]

**Review by apavalache — APPROVED** [review]

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsViewController.swift [hunk 7]
```diff
@@ -43,13 +71,14 @@ class TestsViewController: UIViewController {
...
+    // MARK: - Appearance
+    private func setAppearance() {
         view.backgroundColor = .white
-        navAppearence()
+        navAppearance()
     }
-    
-    private func navAppearence() {
+
+    private func navAppearance() {
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsViewController.swift:81]: better naming? nav appearance doesn't say nothing... better have a read again here about naming of methods: https://github.com/raywenderlich/swift-style-guide

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsViewController.swift [hunk 8]
```diff
@@ -72,31 +101,95 @@ class TestsViewController: UIViewController {
...
+                withIdentifier: ChaptersTableViewCell.Details.identifier, for: indexPath)
+            as? ChaptersTableViewCell else { return UITableViewCell() }
+            cell.updateCellViewforPassed(/* PassedModel @ indexPath */)
+            return cell
+        }
+    }
+
+    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+        switch self.currentButton {
+        case NavButtons.chaptersButton:
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsViewController.swift:181]: you can have case for final and a default, or case with chapters and passed button and final :)

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsViewController.swift [hunk 9]
```diff
@@ -72,31 +101,95 @@ class TestsViewController: UIViewController {
...
+            // height + top
+            // 90     + 25
+            return 115
+        case .passedButton:
+            // height + top
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsViewController.swift:186]: I don't like the comments :)

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsViewController.swift [hunk 10]
```diff
@@ -72,31 +101,95 @@ class TestsViewController: UIViewController {
...
+            guard let cell = tableView.dequeueReusableCell(
+                withIdentifier: ChaptersTableViewCell.Details.identifier, for: indexPath)
+            as? ChaptersTableViewCell else { return UITableViewCell() }
+            cell.updateCellViewforPassed(/* PassedModel @ indexPath */)
+            return cell
+        }
+    }
+
+    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+        switch self.currentButton {
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsViewController.swift:180]: do we need self?

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsViewController.swift [hunk 11]
```diff
@@ -72,31 +101,95 @@ class TestsViewController: UIViewController {
...
+    private func getPassed() -> [String] {
+        return ["1", "2", "3"]
+    }
+}
+
+extension TestsViewController: UITableViewDataSource, UITableViewDelegate {
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        switch self.currentButton {
+        case .chaptersButton:
+            return self.chaptersDataSource.count
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsViewController.swift:146]: do we need self?

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsViewController.swift [hunk 12]
```diff
@@ -72,31 +101,95 @@ class TestsViewController: UIViewController {
...
-    
+
     private func makeButtonSelected() {
         switch currentButton {
+        case .chaptersButton:
+            setSelectedButtonAppearance(button: chaptersButton)
+            setNormalButtonAppearance(button: finalButton)
+            setNormalButtonAppearance(button: passedButton)
+        case .finalButton:
+            setNormalButtonAppearance(button: chaptersButton)
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsViewController.swift:122]: for the future, ignore for this code review: I would create a custom button, with a custom state. For example isSelected and we can setup there the button color, tint color, etc. and here just to select the button which is selected.

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsViewController.swift [hunk 13]
```diff
@@ -7,30 +7,58 @@
...
+        testsTableView.dataSource = self
+        testsTableView.delegate = self
+        // remove cell separator
+        testsTableView.separatorColor = .clear
     }
-    
+
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(true)
+        chaptersDataSource = self.getChapters()
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsViewController.swift:51]: again with self

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsViewController.swift [hunk 14]
```diff
@@ -7,30 +7,58 @@
...
+    var chaptersDataSource: [String] = []
+    var passedDataSource: [String] = []
+    let finalDataSource: [FinalTestModel] = [
+        FinalTestModel(title: "Generate final random test"),
+        FinalTestModel(title: "Scan the QR code and start the coresponding test")
+    ]
+
     override func viewDidLoad() {
         super.viewDidLoad()
+        // register cells Nib's
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/TestsSegmentedController/TestsViewController.swift:35]: on iPhone 8 the second view si cut, would be nice to be smaller or to have both on the screen

Simulator Screen Shot - iPhone 8 - 2022-08-09 at 13 45 56

**Pull request merged**